plugins/secureboot-paper/src/main/java/net/civmc/secureboot/paper/SecureBootPlugin.java

```diff
@@ -0,0 +1,132 @@
+package net.civmc.secureboot.paper;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.bukkit.Bukkit;
+import org.bukkit.OfflinePlayer;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.server.PluginDisableEvent;
+import org.bukkit.event.server.PluginEnableEvent;
+import org.bukkit.event.server.ServerLoadEvent;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.JavaPlugin;
+import org.jetbrains.annotations.NotNull;
+import java.util.Collections;
+import java.util.Map;
+import java.util.TreeMap;
+
+enum PluginStatus {
+    AWAITING_STATUS,
+    LOADED,
+    ENABLED
+}
+
+@SuppressWarnings("UnstableApiUsage")
+public final class SecureBootPlugin extends JavaPlugin implements Listener {
+    private final Map<String, PluginStatus> pluginState = Collections.synchronizedMap(new TreeMap<>(String.CASE_INSENSITIVE_ORDER));
+    private final MiniMessage mm = MiniMessage.miniMessage();
+    private @NotNull Component kickMessage = Component.text("Server is in maintenance mode.");
+    private boolean acceptLogins = false;
+
+    @Override
+    public void onLoad() {
+        saveDefaultConfig();
+        // ensure required plugins have a status
+        getConfig().getStringList("required_plugins").forEach((plugin) -> this.pluginState.put(plugin, PluginStatus.AWAITING_STATUS));
+        getSLF4JLogger().debug("Awaiting status for: {}", String.join(", ", this.pluginState.keySet()));
+
+        // set kick message from config
+        String rawKickMessage = getConfig().getString("kick_message", "Server is in maintenance mode.");
+        kickMessage = mm.deserialize(rawKickMessage);
+    }
+
+    @Override
+    public void onEnable() {
+        Bukkit.getPluginManager().registerEvents(this, this);
+
+        // add all plugins loaded by the server
+        for (final Plugin plugin : Bukkit.getPluginManager().getPlugins()) {
+            this.pluginState.replace(
+                plugin.getPluginMeta().getName(),
+                PluginStatus.AWAITING_STATUS,
+                // Just in case this plugin was enabled prior to SecureBoot
```

Comment:
Is this actually possible? Remove if not

Reply:
While the plugin is set to load fairly early, I don't know of anyway to guarantee the plugins its monitoring load after it. The best I can think of is to strictly depend on them in the plugin config and set the load order that way. Even then though, paper still may decide to do its own thing. It's a good sanity check imo especially because we do not physically  control plugin loading.

Reply:
> Is this actually possible? Remove if not

Yes. [relevant](https://github.com/CivMC/Civ/pull/778#issuecomment-2689257818)